### PR TITLE
Add getLocalFileSystem API and enforce non-null config to getFileSystem

### PR DESCRIPTION
**Pull request by majetideepak**
As discussed here https://github.com/facebookincubator/velox/discussions/8071, enforce a non-null configuration is passed to the `getFileSystem` API call.
Add `getLocalFileSystem` API and use it where a local file is expected to be used.

### CODE REVIEW

**Comment by majetideepak** [comment]
@mbasmanova, @xiaoxmeng Can you provide feedback on this? Thanks!

File: velox/connectors/hive/FileHandle.cpp [hunk 1]
```diff
@@ -43,8 +43,13 @@ std::shared_ptr<FileHandle> FileHandleGenerator::operator()(
   {
     MicrosecondTimer timer(&elapsedTimeUs);
     fileHandle = std::make_shared<FileHandle>();
-    fileHandle->file = filesystems::getFileSystem(filename, properties_)
-                           ->openFileForRead(filename);
+    if (properties_) {
+      fileHandle->file = filesystems::getFileSystem(filename, properties_)
+                             ->openFileForRead(filename);
+    } else {
+      fileHandle->file =
```
Comment (xiaoxmeng) [inline velox/connectors/hive/FileHandle.cpp:50]: Why we assume local file system access if properties_ is not set?

File: velox/exec/SpillFile.cpp [hunk 2]
```diff
@@ -63,7 +63,7 @@ SpillWriteFile::SpillWriteFile(
     const std::string& pathPrefix,
     const std::string& fileCreateConfig)
     : id_(id), path_(fmt::format("{}-{}", pathPrefix, ordinalCounter_++)) {
-  auto fs = filesystems::getFileSystem(path_, nullptr);
+  auto fs = filesystems::getLocalFileSystem();
```
Comment (xiaoxmeng) [inline velox/exec/SpillFile.cpp:66]: We can't change like this. In Meta, we spill to remote storage.

**Review by xiaoxmeng — COMMENTED** [review]
@majetideepak thanks for the change. We can't assume spill always go to local storage. I am wondering can we pass the per-query session properties to per-file options? Why we have to create one filesystem instance per each session properties setting? The filesystem is just an software abstraction. We just need a path prefix (or file scheme) to dispatch the fs call to a particular fs module to serve the request. Internally, it can have a cache with one instance per each config setting (which is specified by per-query session properties) which could be transparent to the user. The config passed to file system instance could be common to all the internal file system instances. Given that, we can separate the file system instance initialization and get. WDYT?

File: velox/connectors/hive/FileHandle.cpp (hunk 1 above)
Comment (majetideepak) [reply velox/connectors/hive/FileHandle.cpp:50]: The underlying assumption is that any remote storage would need non-null config. If properties are missing, then the local file system is used.

Comment (majetideepak) [reply velox/connectors/hive/FileHandle.cpp:50]: I should be adding a check for the file scheme here for the local file system.

File: velox/exec/SpillFile.cpp (hunk 2 above)
Comment (majetideepak) [reply velox/exec/SpillFile.cpp:66]: What scheme does your remote storage use? Does your remote storage not need any properties?

**Comment by majetideepak** [comment]
>  I am wondering can we pass the per-query session properties to per-file options? Why we have to create one filesystem instance per each session properties setting? The filesystem is just an software abstraction.  We just need a path prefix (or file scheme) to dispatch the fs call to a particular fs module to serve the request .....

Based on your explanation, I am guessing your file scheme has the endpoint and other config encoded.
I also assume each user specifies the credentials/config via session properties.

> The config passed to file system instance could be common to all the internal file system instances
This is not possible for S3 where the endpoint defines a filesystem. So there has to be a separation at the file system level.

Can you confirm how the file scheme, endpoint, credentials work for the Meta's file system?

**Pull request closed without merging**